### DISCUSSION TITLE
`zsock_listen` errors out with `134`

### DISCUSSION THREAD

**Discussion by Dimfred** [zephyrproject-rtos/zephyr#78315] [General] [answered]
Hi,
I am trying to run an http server on my stm32 nucleo_u575zi_q.
I am using an ESP8266 wifi module and have created a board overlay file for this.
I have verified that wifi works, since I am able to perform an `http_get` to a remote server.

Currently I am trying to create an http server on the stm32, using the examples provided by you guys, but I run here into issues.
Problem is when `zsock_listen` is called it errors out with `errno=134` (by the zephyr docs thats `ENOTSUP` aka not supported?).
Any suggestions would be appreaciated, I am attaching a repo with a minimally verifiable example if someone could take a look I would appreciate that a lot.

```
*** Booting Zephyr OS build v3.7.0-2509-g6654c2a2f629 ***
[00:00:03.788,000] <inf> net_config: Initializing network
[00:00:03.794,000] <inf> net_config: Waiting interface 1 (0x20000f58) to be up...
[00:00:04.803,000] <inf> net_config: Running dhcpv4 client...
[00:00:04.809,000] <err> net_config: Timeout while waiting network interface
[00:00:04.816,000] <err> net_config: Network initialization failed (-115)
[00:00:04.824,000] <dbg> net_sock: zsock_socket_internal: (0x20002ec8): socket: ctx=0x20003318, fd=3
listen: 134
[00:00:04.835,000] <dbg> net_sock: z_impl_zsock_close: (0x20002ec8): close: ctx=0x20003318, fd=3
```

Idk why it prints the network init failed, but despite this error I could still write sockets, aka do my http_get, without issues.

https://github.com/Dimfred/zephyr-minimal-example-zsock-listen
In `config.h` I changed the ssid/password.
It might also start if you don't have access to any network maybe? But not sure here.

Zephyr is @commit: 6654c2a2f6298d3395770291626edbf3c7d3ff65

Any help is really, really appreciated, I am starting to loose my mind.

Cheers
DimfreD

**Comment by jukkar** [comment]
I just tried your example application. I had to do some changes in order to compile things, not sure if these are relevant to anything as I do not use C++ normally but without those I got some compile issues.
Anyway, the thing worked just fine for me when using native-sim
```
$ build/.../zephyr/zephyr.exe -attach_uart
WARNING: Using a test - not safe - entropy source
uart connected to pseudotty: /dev/pts/3
*** Booting Zephyr OS build v3.7.0-2619-g95599968d907 ***
[00:00:00.000,000] <inf> net_config: Initializing network
[00:00:00.000,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:00.000,000] <inf> net_config: Running dhcpv4 client...
[00:00:00.000,000] <dbg> net_sock: zsock_socket_internal: (main): socket: ctx=0x808a8e0, fd=3
[00:00:00.000,000] <dbg> net_ctx: net_context_bind: (main): Context 0x808a8e0 binding to TCP 0.0.0.0:80 iface 1 (0x8088790)
Main thread done, sleeping forever...[00:00:12.780,000] <dbg> net_ctx: net_context_bind: (rx_q[0]): Context 0x808a97c binding to TCP 192.0.2.1:35046 iface 1 (0x8088790)
[00:00:12.840,000] <dbg> net_sock: zsock_accepted_cb: (rx_q[0]): parent=0x808a8e0, ctx=0x808a97c, st=0
[00:00:14.280,000] <dbg> net_sock: zsock_received_cb: (rx_q[0]): ctx=0x808a97c, pkt=0x80929a0, st=0, user_data=0
[00:00:15.540,000] <dbg> net_sock: zsock_received_cb: (rx_q[0]): ctx=0x808a97c, pkt=0x8092960, st=0, user_data=0
[00:00:16.800,000] <dbg> net_sock: zsock_received_cb: (rx_q[0]): ctx=0x808a97c, pkt=0x80929e0, st=0, user_data=0
[00:00:19.020,000] <dbg> net_sock: zsock_received_cb: (rx_q[0]): ctx=0x808a97c, pkt=0, st=0, user_data=0
[00:00:19.020,000] <dbg> net_sock: zsock_received_cb: (rx_q[0]): Set EOF flag on pkt 0x80929e0
^C
Stopped at 20.740s
```
The diff is here:
```
diff --git a/app/prj.conf b/app/prj.conf
index e17a897..475f35a 100644
--- a/app/prj.conf
+++ b/app/prj.conf
@@ -1,5 +1,6 @@
+CONFIG_CPP=y
 CONFIG_CPP_EXCEPTIONS=y
-CONFIG_GLIBCXX_LIBCPP=y
+#CONFIG_GLIBCXX_LIBCPP=y
 #
 # LOGGING
 CONFIG_NET_LOG=y
@@ -27,7 +28,7 @@ CONFIG_NET_DHCPV4=y
 
 # HOME CONFIG
 # CONFIG_NET_DHCPV4=n
-# CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.168.178.111"
+CONFIG_NET_CONFIG_MY_IPV4_ADDR="192.0.2.1"
 # CONFIG_NET_CONFIG_MY_IPV4_GW="192.168.178.1"
 
 # EXTERNAL ROUTER CONFIG
diff --git a/app/src/main.cpp b/app/src/main.cpp
index f47e007..37dbd87 100644
--- a/app/src/main.cpp
+++ b/app/src/main.cpp
@@ -1,5 +1,5 @@
 #include <errno.h>
-#include <exception>
+//#include <exception>
 #include <stdlib.h>
 
 #include <zephyr/device.h>
@@ -18,7 +18,7 @@
 /// MAIN
 int run_main()
 {
-    try
+  //  try
     {
         // wifi_init();
         // wifi_connect(SSID, PSK);
@@ -59,11 +59,11 @@ int run_main()
             return 1;
         }
     }
-    catch (std::exception &e)
-    {
-        printk("Exception: %s\n", e.what());
-        return 1;
-    }
+    // catch (std::exception &e)
+    //{
+    //    printk("Exception: %s\n", e.what());
+    //    return 1;
+    //}
 
     printk("Main thread done, sleeping forever...");
     k_sleep(K_FOREVER);
```
Compilation complains so I am missing some config options:
```
$ west build -p -b native_sim -d ../build/listen-enotsup-error zephyr-minimal-example-zsock-listen/app -- \
-DCONFIG_NET_SOCKETS_LOG_LEVEL_DBG=y \
-DCONFIG_NET_CONTEXT_LOG_LEVEL_DBG=y \
-DCONFIG_NATIVE_UART_AUTOATTACH_DEFAULT_CMD="\"gnome-terminal -- screen %s\"" \
-DCONFIG_NET_SHELL=y
-- west build: making build dir /.../zephyrproject/build/listen-enotsup-error pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /.../zephyrproject/zephyr/zephyr-minimal-example-zsock-listen/app
-- CMake version: 3.22.1
-- Found Python3: /usr/bin/python3 (found suitable version "3.10.12", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /.../.cache/zephyr
-- Zephyr version: 3.7.99 (/.../zephyrproject/zephyr)
-- Found west (found suitable version "1.1.0", minimum required is "0.14.0")
-- Board: native_sim, qualifiers: native
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.8 (/.../zephyr-sdk-0.16.8)
-- Found toolchain: host (gcc/ld)
-- Found Dtc: /.../zephyr-sdk-0.16.8/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /.../zephyrproject/zephyr/boards/native/native_sim/native_sim.dts
-- Generated zephyr.dts: /.../zephyrproject/build/listen-enotsup-error/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /.../zephyrproject/build/listen-enotsup-error/zephyr/include/generated/zephyr/devicetree_generated.h
-- Including generated dts.cmake file: /.../zephyrproject/build/listen-enotsup-error/zephyr/dts.cmake

warning: CPP_EXCEPTIONS (defined at lib/cpp/Kconfig:133) was assigned the value 'y' but got the
value 'n'. Check these unsatisfied dependencies: (!MINIMAL_LIBCPP) (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_CPP_EXCEPTIONS and/or look up
CPP_EXCEPTIONS in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.


warning: PRINTK_BUFFER_SIZE (defined at subsys/debug/Kconfig:229) was assigned the value '2048' but
got the value ''. Check these unsatisfied dependencies: USERSPACE (=n). See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_PRINTK_BUFFER_SIZE and/or look up
PRINTK_BUFFER_SIZE in the menuconfig/guiconfig interface. The Application Development Primer,
Setting Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be
helpful too.


warning: The choice symbol NEWLIB_LIBC (defined at lib/libc/Kconfig:95) was selected (set =y), but
PICOLIBC (defined at lib/libc/Kconfig:81) ended up as the choice selection. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_NEWLIB_LIBC and/or look up NEWLIB_LIBC in
the menuconfig/guiconfig interface. The Application Development Primer, Setting Configuration
Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful too.
...
```

**Reply by Dimfred** [reply]
Thanks for trying. Yeah I adapted my example, but I receive still the same error. I mean the native sim does something different I guess and on my hardware it does not run. Do you have some hardware to try on? Or what exactly does the native sim do?

**Reply by jukkar** [reply]
native-sim is just a board where the application is run as a Linux application. The application run in native-sim should work the same way as in any other board, except that time advances immediately by default. I have suitable hw I could try to run this, I will report here what is the outcome.

**Comment by jukkar** [comment]
I have a nucleo_f767zi board that I tried
```
west build -p -b nucleo_f767zi/stm32f767xx -d ../build/listen-enotsup-error zephyr-minimal-example-zsock-listen/app -- \
-DCONFIG_NET_SOCKETS_LOG_LEVEL_DBG=y \
-DCONFIG_NET_CONTEXT_LOG_LEVEL_DBG=y \
-DCONFIG_NET_SHELL=y \
-DCONFIG_NET_MGMT_EVENT_STACK_SIZE=2048
```
I am not seeing any issue with listen.
```
[00:00:01.012,000] <inf> net_config: IPv4 address: 192.0.2.1
[00:00:01.020,000] <inf> net_config: Running dhcpv4 client...
[00:00:01.029,000] <err> net_config: Timeout while waiting network interface
[00:00:01.038,000] <err> net_config: Network initialization failed (-115)
[00:00:01.048,000] <dbg> net_sock: zsock_socket_internal: (main): socket: ctx=0x20023ac0, fd=3
[00:00:01.059,000] <dbg> net_ctx: net_context_bind: (main): Context 0x20023ac0 binding to TCP 0.0.0.0:80 iface 1 (0x20021890)
Main thread done, sleeping forever...[00:00:01.879,000] <inf> net_dhcpv4: Received: 192.168.88.36
[00:00:01.887,000] <inf> net_config: IPv4 address: 192.168.88.36
[00:00:01.896,000] <inf> net_config: Lease time: 600 seconds
[00:00:01.904,000] <inf> net_config: Subnet: 255.255.255.0
[00:00:01.912,000] <inf> net_config: Router: 192.168.88.1
```

**Reply by Dimfred** [reply]
okay interesting. So just checked your board, you have one with an RJ45 port and it is stated as ethernet compliant, which board perse is not. Do you think this could be the issue maybe?

**Reply by jukkar** [reply]
I have used this board for several years for network testing, the Ethernet is working with it fine.
Not sure if what could be your issue with `listen()` call, I tried to see from the sources where it could return `ENOTSUP`, but did not spot anything.

**Reply by jukkar** [reply]
The network initialization errors look a bit bad, you could turn off config library to see does it change anything. So `CONFIG_NET_CONFIG_SETTINGS=n`

**Reply by Dimfred** [reply]
didn't help sadly still the same error. Also you say you have used `this` board for network testing. Are you talking about my board `u575zi_q` or about your `f767zi`.

**Reply by jukkar** [reply]
> Are you talking about my board `u575zi_q` or about your `f767zi`.

I am talking about my f767zi board.

**Comment by jukkar** [comment]
One thing you could try to do, is to figure out where that `ENOTSUP` code is coming from, perhaps we can then try to get to the root cause of the issue.

**Reply by Dimfred** [reply]
Yeah I guess I have to go further down then. I'll try and report back.

**Comment by jukkar** [comment]
One thing I just realized, you seem to have wifi so it must be offloaded one (right?). In this case your issue is in the offloaded driver / implementation and is not really a zephyr issue.

**Reply by Dimfred** [reply]
Yes I am using an ESP8266 wifi dongle sry completely forgot to mention that. (EDIT: ah actually mentioned this)

**Reply by Dimfred** [reply]
zephyr/subsys/net/ip/net_context.c:1092
![image](https://github.com/user-attachments/assets/85a52cb6-8902-4482-9684-5818c6c3453c)

 `zephyr/drivers/wifi/esp_at/esp_offload.c`
![image](https://github.com/user-attachments/assets/3a56d8bc-c415-46a7-9fb3-5975b7201f90)

yeah.... okay should've went deeper, didn't event consider that... (am super new to embedded still not getting stuff :D)

Question then for me remains, this is a `zephyr/drivers/wifi/esp_at/esp_offload.c`. Is this a zephyr issue, and zephyr did not implement the functionality, or is it just not implemented, because the hardware itself really doesn't support it.

I would actually either now order a device where it is supported, or maybe try to implement listen on my own. Is that a good idea?

**Comment by jukkar** [comment, marked as answer]
Perhaps it is this piece of code then that is returning the ENOTSUP

https://github.com/zephyrproject-rtos/zephyr/blob/b211e1b3a35e32e8af48e7118a5cf114b1ce5b25/drivers/wifi/esp_at/esp_offload.c#L25

**Reply by Dimfred** [reply]
jupp it is figured it out a little after you...^^

**Reply by Dimfred** [reply]
Okay I now found [the issue](https://github.com/zephyrproject-rtos/zephyr/issues/51642) for this. Guess I will have to try to find another way until this is implemented.
@jukkar thank you very much for your support, I owe you a beer :3
